Added "How to work with contracts and transactions in Java". (#21237)
digital-asset/daml: sdk/docs/manually-written/sdk/sdlc-howtos/applications/develop/contracts-and-transactions-in-java.rst

.. Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _howto-applications-work-with-contracts-java:

How to work with contracts and transactions in Java
===================================================

> When writing applications for the Canton Ledger in Java, it is convenient to work with a representation of Daml templates and data

Reviewer: ```suggestion
When writing Canton Network applications in Java, it is convenient to work with a representation of Daml templates and data
```

types in Java that closely resemble the original Daml code while still being as close to the native types of Java as
possible. To achieve this, use the :ref:`Daml Codegen for Java <daml-codegen-java>` to generate Java types based on a Daml
model. You can then use these types in your Java code when reading information from and sending data to the ledger.

Setup the codegen
-----------------

Run and configure the code generator for Java according to :ref:`Daml Codegen for Java <daml-codegen-java>` to use and
generate the Java classes for your project.

See also in :ref:`Generated code <daml-codegen-java-generated-code>` how the generated code looks for the Daml built-in
and user-defined types.

Use the generated classes in your project
-----------------------------------------

> In order to compile the resulting Java classes, you need to add the Java bindings library <TODO: add link to Java bindings>

Reviewer: use `:brokenref:`

Reviewer: here and in all other cases

The Java bindings library lets you connect to a ledger via a secure connection. To do so, invoke ``withSslContext`` and
pass an ``SslContext``.

.. warning::

  Use the default plaintext connection only when connecting to a locally running ledger for development purposes.

Secure connections to a ledger must be configured to use client authentication certificates, which can be provided by a ledger operator.

For information on how to set up an ``SslContext`` with the provided certificates for client authentication, please consult the gRPC documentation on
`TLS with OpenSSL <https://github.com/grpc/grpc-java/blob/master/SECURITY.md#tls-with-openssl>`_ as well as the
`HelloWorldClientTls <https://github.com/grpc/grpc-java/blob/70b1b1696a258ffe042c7124217e3a7894821444/examples/src/main/java/io/grpc/examples/helloworldtls/HelloWorldClientTls.java#L46-L57>`_ example of the ``grpc-java`` project.

Advanced connection settings
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Sometimes the default settings for gRPC connections or channels are not suitable. For such situations, create a custom
`NettyChannelBuilder <https://grpc.github.io/grpc-java/javadoc/io/grpc/netty/NettyChannelBuilder.html>`_ object and

> pass the it to the ``newBuilder`` static method defined over `DamlLedgerClient <javadocs/com/daml/ledger/rxjava/DamlLedgerClient.html>`_.

Author: Is this link safe? @pawelulewicz-da 
The same already existed in https://github.com/digital-asset/daml/blob/2ca992544ab030abdaabe7ec11a0b5d750562493/sdk/docs/manually-written/sdk/sdlc-howtos/applications/integrate/bindings-java/index.rst?plain=1#L59.

Author: when locally deploying the daml docs is broken ofcourse

..  TODO                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ fix me

Alternative ways to access the Ledger API
-----------------------------------------

The Java bindings client library is not the only way to access the Ledger API. You can also use the gRPC bindings or OpenAPI
bindings directly.

Use gRPC bindings
^^^^^^^^^^^^^^^^^

For each gRPC endpoint that you want to call from your Java application, create a gRPC `StreamObserver <https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html>`_ providing
implementations of the onNext, onError and onComplete observer methods.
To decode and encode the gRPC messages, use the fromProto and toProto methods of the generated classes of the Java
bindings library. <TODO(#370): add link to Java bindings library fromProto and toProto methods>

Use OpenAPI definitions
^^^^^^^^^^^^^^^^^^^^^^^

The OpenAPI definitions provide a definition for each Ledger API service and allows you to access it via the JSON
Ledger API. Use those definitions to encode/decode the gRPC messages as/from JSON payloads required by the JSON Ledger

> API. For more details see <Get started with Canton and the JSON Ledger API TODO(#369): add link to it>.

Author: @jarekr-da could u check the OpenAPI part?
